Repalce usage of legacy vars <Package>_[LIB|TEST]_[REQUIRED|OPTIONAL]_DEP_[PACKAGES|TPLS] (#63)

bartlettroscoe opened the pull request:
### Description

The primary purpose of this PR is remove the setting and usage of the legacy vars `<Package>_[LIB|TEST]_[REQUIRED|OPTIONAL]_DEP_[PACKAGES|TPLS] ` that treat external TPLs and internal Packages as separate entities and instead use the new vars:

* `<Package>_[LIB|TEST]_[DEFINED|ENABLED]_DEPENDENCIES`
* `<Package>_[LIB|TEST]_DEP_REQUIRED_<depPkg>`
* `<Package>_PACKAGE_BUILD_STATUS`

It also replaces some of the loops over just internal packages and TPLs with the entire list of packages (and then switching on  `<Package>_PACKAGE_BUILD_STATUS`).

This takes us one step closer to completing #63 and being able to allow packages to be defined internally or externally in a fluid way.

There are still some more refactoring needed to complete the initial use case driving #63 ([Use Case 3: Configure/build pointing to a subset of already installed TriBITS packages in same repo](https://docs.google.com/document/d/1T9KKrtgFm8-TSEuq119gdDR9tJMlBN3nAVjcN15Ge6k/edit#heading=h.wm8s8of6cwg9)).

### Instructions for Reviewers

This PR involving large blocks of code around in the same file and between different files.  But these simple movements of code (with no changes in that code) are done in their own commits.  Therefore, **the best way to review this PR is to review it commit-by-commit**.

There were also some other small refactorings and updates that were done and these are isolated into their own commits.

bartlettroscoe commented:
Hello @KyleFromKitware, this is a big PR (see above suggestions for how to do the review commit-by-commit).  It would be great if you could review it before you go on vacation next week.  But if you can't, I will just merge it and set it up for a post-merge review when you get back in Jan 2023.

rabartlett1972 reviewed (approved):
Given that Kyle is out until mid Jan, I will go ahead and allow this to merge and set up for a post-merge review.

bartlettroscoe commented:
Hello @KyleFromKitware, can you please do a post-merge review of this PR once you get back and have some time?

bartlettroscoe commented:
@KyleFromKitware, is this all of the feedback you have on this PR?

bartlettroscoe commented on tribits/core/package_arch/TribitsAdjustPackageEnables.cmake:
```diff
@@ -979,76 +951,64 @@ function(tribits_private_print_disable_required_package_enable
-    if ("${LIBRARY_DEP}" STREQUAL "TRUE")
-
+  tribits_get_package_enable_status(${fwdDepPkgName}  ""  fwdDepPkgEnableVarName)
+  if (${fwdDepPkgEnableVarName} OR ${fwdDepPkgEnableVarName} STREQUAL "")
```
Quotes is not correct in this case and would give the wrong behavior.  The variable `fwdDepPkgEnableVarName` dereferences to the name of an enable var like `<Project>_ENABLE_<Package>` or `TPL_ENABLE_<Package>` (which is a bit of indirectly that we need in logic like this that deals with external and internal packages).  So this if statement is really something like:

```
 if (<Project>_ENABLE_<Package> OR <Project>_ENABLE_<Package> STREQUAL "")
```

or:

```
 if (TPL_ENABLE_<Package> OR TPL_ENABLE_<Package> STREQUAL "")
```

But to make this more clear, I will change the last part to:

```
"${${fwdDepPkgEnableVarName}}" STREQUAL ""
```

KyleFromKitware replied:
I see, thanks for the clarification.

KyleFromKitware commented:
This is it for the moment, but I'll have more time on Wednesday and Thursday if you'd like me to take another look after our biweekly meeting tomorrow.

bartlettroscoe commented:
> This is it for the moment, but I'll have more time on Wednesday and Thursday if you'd like me to take another look after our biweekly meeting tomorrow.

@KyleFromKitware, sure, this is a lot of code touched and refactored so it should take a couple of hours to go over all of it carefully.

For now I will just post a PR that addresses the comments [above](https://github.com/TriBITSPub/TriBITS/pull/546#pullrequestreview-1212269205) and then I can post another one after that to address any other comments.

bartlettroscoe commented on tribits/ci_support/TribitsWriteXmlDependenciesFiles.cmake:
```diff
@@ -197,3 +200,53 @@ function(tribits_write_deps_to_xml_string  packageName  listType  xmlVarInOut)
+      set(matchesRequriedOrOptional TRUE)
+    endif()
+
+    SET(matchesPackagesOrTpls FALSE)
```
Addressed in PR https://github.com/TriBITSPub/TriBITS/pull/550

bartlettroscoe commented on tribits/ci_support/TribitsWriteXmlDependenciesFiles.cmake:
```diff
@@ -197,3 +200,53 @@ function(tribits_write_deps_to_xml_string  packageName  listType  xmlVarInOut)
+        ((packagesOrTpls STREQUAL "TPLS")
+        AND (${depPkg}_PACKAGE_BUILD_STATUS STREQUAL "EXTERNAL"))
+      )
+      SET(matchesPackagesOrTpls TRUE)
```
Addressed in PR https://github.com/TriBITSPub/TriBITS/pull/550
